cmd/kubeadm/app/phases/addons/dns/dns.go

```diff
@@ -127,6 +131,161 @@ func createKubeDNSAddon(deploymentBytes, serviceBytes []byte, client clientset.I
+	if err := apiclient.CreateOrUpdateConfigMap(client, coreDNSConfigMap); err != nil {
+		return err
+	}
+
+	coreDNSClusterRoles := &v1beta1.ClusterRole{}
+	if err := kuberuntime.DecodeInto(api.Codecs.UniversalDecoder(), clusterroleBytes, coreDNSClusterRoles); err != nil {
+		return fmt.Errorf("unable to decode CoreDNS clusterroles %v", err)
+	}
+
+	// Create the Clusterroles for CoreDNS or update it in case it already exists
+	if err := apiclient.CreateOrUpdateClusterRole(client, coreDNSClusterRoles); err != nil {
+		return err
+	}
+
+	coreDNSClusterRolesBinding := &v1beta1.ClusterRoleBinding{}
+	if err := kuberuntime.DecodeInto(api.Codecs.UniversalDecoder(), clusterrolebindingBytes, coreDNSClusterRolesBinding); err != nil {
+		return fmt.Errorf("unable to decode CoreDNS clusterrolebindings %v", err)
+	}
+
+	// Create the Clusterrolebindings for CoreDNS or update it in case it already exists
+	if err := apiclient.CreateOrUpdateClusterRoleBinding(client, coreDNSClusterRolesBinding); err != nil {
+		return err
+	}
+
+	coreDNSServiceAccount := &v1.ServiceAccount{}
+	if err := kuberuntime.DecodeInto(api.Codecs.UniversalDecoder(), serviceAccountBytes, coreDNSServiceAccount); err != nil {
+		return fmt.Errorf("unable to decode CoreDNS configmap %v", err)
+	}
+
+	// Create the ConfigMap for CoreDNS or update it in case it already exists
+	if err := apiclient.CreateOrUpdateServiceAccount(client, coreDNSServiceAccount); err != nil {
+		return err
+	}
+
+	coreDNSDeployment := &extensions.Deployment{}
+	if err := kuberuntime.DecodeInto(api.Codecs.UniversalDecoder(), deploymentBytes, coreDNSDeployment); err != nil {
+		return fmt.Errorf("unable to decode CoreDNS deployment %v", err)
+	}
+
+	// Create the Deployment for CoreDNS or update it in case it already exists
+	if err := apiclient.CreateOrUpdateDeployment(client, coreDNSDeployment); err != nil {
+		return err
+	}
+
+	coreDNSService := &v1.Service{}
+	if err := kuberuntime.DecodeInto(api.Codecs.UniversalDecoder(), serviceBytes, coreDNSService); err != nil {
+		return fmt.Errorf("unable to decode coreDNS service %v", err)
+	}
+
+	// Can't use a generic apiclient helper func here as we have to tolerate more than AlreadyExists.
+	if _, err := client.CoreV1().Services(metav1.NamespaceSystem).Create(coreDNSService); err != nil {
+		// Ignore if the Service is invalid with this error message:
+		// 	Service "CoreDNS" is invalid: spec.clusterIP: Invalid value: "10.96.0.10": provided IP is already allocated
+
+		if !apierrors.IsAlreadyExists(err) && !apierrors.IsInvalid(err) {
+			return fmt.Errorf("unable to create a new CoreDNS service: %v", err)
+		}
+
+		if _, err := client.CoreV1().Services(metav1.NamespaceSystem).Update(coreDNSService); err != nil {
+			return fmt.Errorf("unable to create/update the CoreDNS service: %v", err)
+		}
+	}
+	return nil
+}
+
+//convSubnet fetches the servicecidr and modifies the mask to the nearest class
```

Comment:
Default is a /12 so it would error out by default :wink:

This is a general DNS problem. We're looking at some solutions here https://github.com/coredns/coredns/issues/1074

kube-dns today captures ALL PTRs. Leading to things like https://github.com/kubernetes/dns/issues/124 and (I think) the ability to hijack the PTR of any IP in the world.

Even if we nail it down to just the service CIDR, we still have problems with PTR and manually added endpoints in K8s, especially in a multi-tenant deployment.